src/precice/config/ParticipantConfiguration.cpp

```diff
@@ -452,73 +465,111 @@ void ParticipantConfiguration::finishParticipantConfiguration(
+      }
     }
 
+    // Now we create the mappingContext, which will be stored permanently
     precice::impl::MappingContext mappingContext;
-    mappingContext.fromMeshID = fromMeshID;
-    mappingContext.toMeshID   = toMeshID;
+    // Copy over data from MappingConfiguration
+    // 1. the mesh data
+    mappingContext.fromMeshID = confMapping.fromMesh->getID();
+    mappingContext.toMeshID   = confMapping.toMesh->getID();
 
+    // Upon creation, the mapping should be empty
     mapping::PtrMapping &map = mappingContext.mapping;
     PRECICE_ASSERT(map.get() == nullptr);
+    // 2. ... and the mappings
     map                                   = confMapping.mapping;
     mappingContext.configuredWithAliasTag = confMapping.configuredWithAliasTag;
 
-    const mesh::PtrMesh &input  = fromMeshContext.mesh;
-    const mesh::PtrMesh &output = toMeshContext.mesh;
+    // Set input and output meshes in the Mapping from the mesh contexts
+    const mesh::PtrMesh &input  = confMapping.fromMesh->isJustInTime() ? confMapping.fromMesh : participant->meshContext(fromMesh).mesh;
+    const mesh::PtrMesh &output = confMapping.toMesh->isJustInTime() ? confMapping.toMesh : participant->meshContext(toMesh).mesh;
     PRECICE_DEBUG("Configure mapping for input={}, output={}", input->getName(), output->getName());
     map->setMeshes(input, output);
 
+    // just-in-time mappings go for now into the participant's mapping context
+    // Add the mapping context to the participant, separated by direction
     if (confMapping.direction == mapping::MappingConfiguration::WRITE) {
       participant->addWriteMappingContext(mappingContext);
     } else {
       PRECICE_ASSERT(confMapping.direction == mapping::MappingConfiguration::READ);
       participant->addReadMappingContext(mappingContext);
     }
 
-    fromMeshContext.meshRequirement = std::max(
-        fromMeshContext.meshRequirement, map->getInputRequirement());
-    toMeshContext.meshRequirement = std::max(
-        toMeshContext.meshRequirement, map->getOutputRequirement());
-
-    fromMeshContext.fromMappingContexts.push_back(mappingContext);
-    toMeshContext.toMappingContexts.push_back(mappingContext);
+    // configure the involved mesh context with connectivity requirements stemming from the mapping
+    // Add the mapping context to the mesh context, only required to later on forward them to the Partition
+    if (!input->isJustInTime()) {
+      participant->meshContext(fromMesh).meshRequirement = std::max(
+          participant->meshContext(fromMesh).meshRequirement, map->getInputRequirement());
+      participant->meshContext(fromMesh).fromMappingContexts.push_back(mappingContext);
+    }
+    if (!output->isJustInTime()) {
+      participant->meshContext(toMesh).toMappingContexts.push_back(mappingContext);
+      participant->meshContext(toMesh).meshRequirement = std::max(
+          participant->meshContext(toMesh).meshRequirement, map->getOutputRequirement());
+    }
```

Comment:
Both `if`-legs could be moved into `MeshContext`. Maybe even the whole block could be moved inside `Mesh`.

Reply:
I am not sure. Then using two functions or what do you have in mind? The two branches are different: input vs output, but also the context we append our own vector to.

Reply:
I mean something in this direction:

```
    if (!input->isJustInTime()) {
      participant->doInputConfiguration(fromMesh, map, mappingContext);
    }
    if (!output->isJustInTime()) {
      participant->doOutputConfiguration(toMesh, map, mappingContext);
    }
```

Maybe some input arguments are not needed. This has two main debefits:

1. the code becomes more readable: you replace the low-level implementation with a high-level function with a descriptive name. 
2. it improves encapsulation: you do not need to access `participant->meshContext(...)` anymore from here but only inside `participant`.

Reply:
Done